File: 1twoSum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        # Empty dictionary
        hashmap = {}
        # Use enumerate to have access to both the index and value from the list
        for i, n in enumerate(nums):
            diff = target - n
            # If the difference is already in the hashmap then you know that the target sum can be reached with
            # the current number and the difference
            if diff in hashmap:
                return [hashmap[diff], i]
            # If the difference is not in the hashmap add n as a key and i as a value
            hashmap[n] = i
        # A return outside the loop is unnecessary since a solution is guaranteed in the list

File: Java/1twoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Initialize HashMap for one pass
        HashMap<Integer, Integer> hashmap = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            // Calculate the difference between the target and the current number
            int diff = target - nums[i];
            // If the difference is already in the HashMap then we can return
            if (hashmap.containsKey(diff)) {
                return new int[]{hashmap.get(diff), i};
            }
            // If the previous check failed then we insert the current number and its index into the HashMap
            hashmap.put(nums[i], i);
        }
        // Return null if nothing is found
        // This should never happen since a solution is guaranteed
        return null;
    }
}